### PR TITLE
Example/anomaly detection example

### PR DESCRIPTION
**Pull request by julien12234**
Contains:
- Jupyter notebook tutorial for the anomaly detection API
- New dataset that represents the number of taxi passengers in New York

### CODE REVIEW

File: examples/XX-anomaly-detection.ipynb [hunk 1]
```diff
@@ -0,0 +1,1355 @@
+{
```
Comment (hrzn) [inline examples/XX-anomaly-detection.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.    # 8 anomalous days
</pre><p>Should be 5 I guess</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/unit8co/darts/pull/1518/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (hrzn) [inline examples/XX-anomaly-detection.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #38.    series_taxi.plot(label="number of taxi passengers", linewidth=0.5, color="#6464ff")
</pre><p>Since there are many time steps, you could maybe make the figure wider here so it's a bit more visible.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/unit8co/darts/pull/1518/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (hrzn) [inline examples/XX-anomaly-detection.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #6.    @interact(selected_anomaly=anomalies_dropdown, delta_plotted_days=delta_plotted_days)
</pre><p>I didn't know this, very cool! Could you also check how it renders once the notebook is built in HTML for the examples documentation page? It's OK if it's not interactive, but would be good if it at least displays well.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/unit8co/darts/pull/1518/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (hrzn) [inline examples/XX-anomaly-detection.ipynb:1]: <p><code>start</code> indicates the fraction of the input series _after which_ the scorers should be trained on. Furthermore I wouldn't say it's best if it's small, but rather explain the tradeoff: small means more data but at the risk of using older parts of the series, whereas large means a smaller but more recent sub-series.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/unit8co/darts/pull/1518/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (hrzn) [inline examples/XX-anomaly-detection.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    anomaly_model.fit(s_taxi_train, start=START)
</pre><p>You can add a <code>;</code> at the end of the line if you want to silence the displayed returned object. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/unit8co/darts/pull/1518/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (hrzn) [inline examples/XX-anomaly-detection.ipynb:1]: <p>You don't seem to be doing anything with <code>anomaly_scores</code> and <code>model_forecasting</code> here. Could perhaps be good to just plot them here?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/unit8co/darts/pull/1518/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (hrzn) [inline examples/XX-anomaly-detection.ipynb:1]: <p>fittable scorer --&gt; fittable detector.</p><p> </p><p>"It requires a parameter <code>high_quantile</code> that indicates the fraction of points considered normal" --&gt; technically this is determined by the combination of <code>high_quantile</code> and <code>low_quantile</code>.</p><p> </p><p>"The detector will flag all timestamps with an anomaly score higher than the 94% quantile. Therefore, the parameter needs to be chosen carefully" --&gt; The "therefore" sounds weird here. It is not the fact that you use the detector with 0.94 that implies that the parameter has to be chosen carefully.</p><p> </p><p>"We are expected to have eight anomalous days, which translates to an estimated 94% of normal days in our dataset. " --&gt; Why 8 ? I thought there were 5 anomalous days. Also: be careful with this, because it's a form of data leakage from the test set (the number of anomalous days is not known in advance). I would feel more confortable if you were using a value such as 90%, 95% or 99% without such a justification, or be really clear that this is just for the sake of the example and in practice the optimal ratio would have to be found.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/unit8co/darts/pull/1518/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (hrzn) [inline examples/XX-anomaly-detection.ipynb:1]: <p>I think a few introductory sentences here (or even above) explaining the conceptual differences between how scorers such as <code>KMeansScorer</code> work -vs- the forecasting based ones would be welcome. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/unit8co/darts/pull/1518/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

Comment (hrzn) [inline examples/XX-anomaly-detection.ipynb:1]: <p>replace width(s) by component(s) :) The width is always fixed and equal to 2 for this time series, as it always has 2 components.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/unit8co/darts/pull/1518/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='47'/>

Comment (hrzn) [inline examples/XX-anomaly-detection.ipynb:1]: <p>by state 1 or 2 I imagine you mean value 1 or value 2?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/unit8co/darts/pull/1518/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='63'/>

**Comment by hrzn** [comment]
@julien12234 that looks very good! I have a couple of minor comments.
Maybe my main recommendation would be to add a short introductory paragraph explaining the difference between the forecasting-based approach and the "direct" approach with `KMeansScorer`, as some people might be wondering.

File: examples/XX-anomaly-detection.ipynb (hunk 1 above)
Comment (julien12234) [reply examples/XX-anomaly-detection.ipynb:1]: Can't i keep state?

Comment (hrzn) [reply examples/XX-anomaly-detection.ipynb]: <p>You could but this is slightly confusing. In general, state != value. The state can be used to enumerate values, for instance you could say:</p><p> </p><ul><li>state 1 corresponds to value 0</li><li>state 2 corresponds to value 1</li><li>state 3 corresponds to value 2</li><li>state 4 corresponds to value 3</li></ul><p> </p><p>Here you seem to be using the words state and value interchangeably. It's fine because this case is very simple, but I think it would be a bit cleaner to use the right terms. </p><p> </p><p>You did it well when you defined the initial synthetic data as follows above:</p><p>"</p><ul><li>State 1: component1=0, component2=1</li><li>State 2: component1=1, component2=0</li></ul><p>"</p>

**Comment by hrzn** [comment]
@julien12234 some of the pictures seem to not be showing, for instance:
image

image

**Comment by D3mianGianoz** [comment]
Hey there @dennisbader, any possibilities of getting this reviewed? :D

**Comment by dennisbader** [comment]
Hi @D3mianGianoz, we're currently finalizing the anomaly detection refactor PR #1477. 
Once it is merged, we can continue with this PR.

**Comment by D3mianGianoz** [comment]
Thanks for the info @dennisbader! Any way I could help or deep knowledge of the core darts library is needed?

File: darts/datasets/__init__.py [hunk 2]
```diff
@@ -491,6 +491,32 @@ def __init__(self):
...
+    References
+    ----------
+    .. [1] https://www.kaggle.com/code/julienjta/nyc-taxi-traffic-analysis
+    """
+
+    def __init__(self):
+        super().__init__(
+            metadata=DatasetLoaderMetadata(
+                "taxi_new_york_passengers.csv",
+                uri=_DEFAULT_PATH + "/taxi_new_york_passengers.csv",
```
Comment (VascoSch92) [inline darts/datasets/__init__.py:511]: this is not a good way to join two strings/paths

use instead
```python
_DEFAULT_PATH / Path('taxi_new_york_passengers.csv')
```

File: darts/datasets/__init__.py [hunk 3]
```diff
@@ -491,6 +491,32 @@ def __init__(self):
...
+
+    References
+    ----------
+    .. [1] https://www.kaggle.com/code/julienjta/nyc-taxi-traffic-analysis
+    """
+
+    def __init__(self):
+        super().__init__(
+            metadata=DatasetLoaderMetadata(
+                "taxi_new_york_passengers.csv",
```
Comment (VascoSch92) [inline darts/datasets/__init__.py:510]: add signature here

File: darts/datasets/__init__.py [hunk 4]
```diff
@@ -491,6 +491,32 @@ def __init__(self):
         )
 
 
+class TaxiNewYorkDataset(DatasetLoaderCSV):
```
Comment (VascoSch92) [inline darts/datasets/__init__.py:494]: Nit: The name of the class should be equal to the name of the dataset, i.e., `TaxiNewYorkPassengersDataset`.

**Comment by dennisbader** [comment]
Closing as implemented in #1477
